test_conformance/integer_ops/test_extended_bit_ops_extract.cpp

```diff
@@ -0,0 +1,278 @@
+//
+// Copyright (c) 2021 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <algorithm>
+#include <numeric>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+#include "procs.h"
+#include "integer_ops_test_info.h"
+#include "harness/testHarness.h"
+
+template <typename T>
+static typename std::make_signed<T>::type
+cpu_bit_extract_signed(T tbase, cl_uint offset, cl_uint count)
+{
+    typedef typename std::make_signed<T>::type signed_t;
+
+    assert(offset <= sizeof(T) * 8);
+    assert(count <= sizeof(T) * 8);
+    assert(offset + count <= sizeof(T) * 8);
+
+    signed_t base = static_cast<signed_t>(tbase);
+    signed_t result;
+
+    if (count == 0)
+    {
+        result = 0;
+    }
+    else if (offset + count < sizeof(T) * 8)
+    {
+        result = base << (sizeof(T) * 8 - offset - count);
+        result = result >> (sizeof(T) * 8 - count);
+    }
+    else
+    {
+        result = base >> offset;
+    }
+
+    return result;
+}
+
+template <typename T>
+static typename std::make_unsigned<T>::type
+cpu_bit_extract_unsigned(T tbase, cl_uint offset, cl_uint count)
+{
+    typedef typename std::make_unsigned<T>::type unsigned_t;
+
+    assert(offset <= sizeof(T) * 8);
+    assert(count <= sizeof(T) * 8);
+    assert(offset + count <= sizeof(T) * 8);
+
+    unsigned_t base = static_cast<unsigned_t>(tbase);
+    unsigned_t result;
+
+    if (count == 0)
+    {
+        result = 0;
+    }
+    else if (offset + count < sizeof(T) * 8)
+    {
+        result = base << (sizeof(T) * 8 - offset - count);
+        result = result >> (sizeof(T) * 8 - count);
+    }
+    else
+    {
+        result = base >> offset;
```

Comment:
If we were to rewrite this as:

```
    if (count == 0)
    {
        result = 0;
    }
    else
    {
        result = base << (sizeof(T) * 8 - offset - count);
        result = result >> (sizeof(T) * 8 - count);
    }
```

and then consider the case where `count != 0` and `!(offset + count == sizeof(T) * 8)` (which would hit the `else` case in the original):

* Given `count != 0`, we know that `sizeof(T) * 8 - offset - count < sizeof(T) * 8`.  The `<<` operation therefore cannot overflow.
* Given `offset + count <= sizeof(T) * 8)`, and given `!(offset + count < sizeof(T) * 8)`, we know `offset + count == sizeof(T) * 8`.
* Given `offset + count == sizeof(T) * 8`, we know `sizeof(T) * 8 - offset - count == 0`.
* `result = base << (sizeof(T) * 8 - offset - count)` would therefore simplify to `result = base << 0`, which can be eliminated.
* Given `offset + count == sizeof(T) * 8`, `sizeof(T) * 8 - count == offset`.
* `base >> (sizeof(T) * 8 - count)` would therefore simplify to `base >> offset`.

In short, this is in fact not a special case, but is an unnecessary substitution of the general case.

Please remove the `else` case, and remove the condition of the `else if`.

Reply:
Fixed - thanks!